src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/Properties/InterceptedProjectProperties/BuildPropertyPage/DefineConstantsValueProvider.cs

```diff
@@ -0,0 +1,37 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using Microsoft.VisualStudio.ProjectSystem.Debug;
+
+namespace Microsoft.VisualStudio.ProjectSystem.Properties
+{
+    /// <summary>
+    /// Supports the setting of constants in the property page, corresponding to the "DefineConstants" property
+    /// </summary>
+    [ExportInterceptingPropertyValueProvider("DefineConstants", ExportInterceptingPropertyValueProviderFile.ProjectFile)]
+    internal sealed class DefineConstantsValueProvider : InterceptingPropertyValueProviderBase
+    {
+        internal const string DefineConstantsPropertyName = "DefineConstants";
+
+        public override Task<string?> OnSetPropertyValueAsync(string propertyName, string unevaluatedPropertyValue, IProjectProperties defaultProperties, IReadOnlyDictionary<string, string>? dimensionalConditions = null)
+        {
+            Dictionary<string, string> constantsDict = AbstractProjectConfigurationDefineConstantsEncoding.ParseIntoDictionary(unevaluatedPropertyValue);
+            return Task.FromResult<string?>(AbstractProjectConfigurationDefineConstantsEncoding.Format(constantsDict));
+        }
+
+        public override Task<string> OnGetEvaluatedPropertyValueAsync(string propertyName, string evaluatedPropertyValue, IProjectProperties defaultProperties)
+        {
+            return OnGetPropertyValueAsync(evaluatedPropertyValue);
+        }
+
+        public override Task<string> OnGetUnevaluatedPropertyValueAsync(string propertyName, string unevaluatedPropertyValue, IProjectProperties defaultProperties)
+        {
+            return OnGetPropertyValueAsync(unevaluatedPropertyValue);
+        }
+
+        private static Task<string> OnGetPropertyValueAsync(string propertyValue)
+        {
```

Comment:
OnGetPropertyValueAsync doesn't need to return a Task, since no parallelizable/threaded work is happening here